src/sync/treiber_stack.rs

```diff
@@ -49,6 +49,20 @@ impl<T> TreiberStack<T> {
         self.try_pop()
     }
 
+    /// Swaps a stack with a new one.  Can be used to take all the
+    /// elements off the stack at once.
+    ///
+    /// Wait-free
+    pub fn swap(&self, new: TreiberStack<T>) -> TreiberStack<T> {
+        let guard = epoch::pin();
+        // new is exclusively owned and so can be written and read
+        // with relaxed atomics
+        let newhead = new.head.swap(None, Relaxed, &guard);
+        let head = self.head.swap_shared(newhead, AcqRel, &guard);
```

Comment:
It may be possible to make this `Relaxed` as well, since there's a data dependency flow between the three calls and `new` is only locally accessible. It's also fine for two adjacent calls to `swap` to get their `self.head.swap_shared`s reordered.